### PR TITLE
fix controller ingestion assignment bug and tests random failures

### PR DESCRIPTION
**Pull request by jshencode**
1.  remove etcd embed server 
2. use gomock and in memory kv.Store implementation for testing
3. fix ingestion assignment bug: add waiting for leader election result before actual assignment loop starts

### CODE REVIEW

File: controller/tasks/etcd/ingestion_assignment.go [hunk 1]
```diff
@@ -173,33 +173,57 @@ func (ia *ingestionAssignmentTask) Done() {
...
+			}
+		case <-ia.stopChan:
+			return
+		}
+
+		ia.logger.With("host", hostName).Infof("elected as leader")
+		namespaces, err := ia.namespaceMutator.ListNamespaces()
 		if err != nil {
-			ia.logger.Fatal(err)
+			ia.logger.Error(err)
```
Comment (shz117) [inline controller/tasks/etcd/ingestion_assignment.go]: should this be fatal as well?

File: controller/tasks/etcd/ingestion_assignment.go [hunk 2]
```diff
@@ -173,33 +173,57 @@ func (ia *ingestionAssignmentTask) Done() {
...
-		case <-tickerChan:
-			if ia.isLeader() {
+			// periodic updates
+			case <-tickerChan:
+				if !ia.isLeader() {
+					ia.logger.With("host", hostName).Infof("host is no longer the leader")
+					break loop
+				}
 				ia.tryRecalculateAllNamespaces()
+			// periodic updates
```
Comment (shz117) [inline controller/tasks/etcd/ingestion_assignment.go]: nit

File: controller/tasks/etcd/ingestion_assignment.go [hunk 3]
```diff
@@ -173,33 +173,57 @@ func (ia *ingestionAssignmentTask) Done() {
...
-		select {
-		// watch changes (i.e: heartbeat changes)
-		case ns := <-ia.watchManager.C():
-			if ia.isLeader() {
+		tickerChan := time.NewTicker(time.Duration(ia.intervalSeconds) * time.Second).C
+
+	loop:
+		for {
+			select {
+			case <- ia.leaderElection.C():
```
Comment (shz117) [inline controller/tasks/etcd/ingestion_assignment.go]: do we need this? this process should already be leader past line 180, right?

Comment (shz117) [reply controller/tasks/etcd/ingestion_assignment.go]: oh, is it for resign?

Comment (jshencode) [reply controller/tasks/etcd/ingestion_assignment.go]: yes, this is for when resign happens

File: controller/tasks/etcd/ingestion_assignment.go (hunk 1 above)
Comment (jshencode) [reply controller/tasks/etcd/ingestion_assignment.go]: done

File: controller/tasks/etcd/ingestion_assignment.go (hunk 2 above)
Comment (jshencode) [reply controller/tasks/etcd/ingestion_assignment.go]: done

**Pull request merged**